src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -188,46 +186,106 @@ private void EnqueueProcessSnapshot(Document newDocument)
-                    _ = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                    currentSyntaxTree = await currentDocument.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
                 }
 
+                // We don't need to grab _lastProcessedDocument in a lock.  We don't care which version of the previous
+                // doc we grab, just that we grab some prior version.  This is only used 
+                var changedSpan = await GetChangedSpanAsync(_lastProcessedDocument, currentDocument, currentSnapshot, cancellationToken).ConfigureAwait(false);
+
+                // Once we're past this point, we're mutating our internal state so we cannot cancel past that.
+                cancellationToken = CancellationToken.None;
+
                 lock (_gate)
                 {
-                    _lastProcessedSnapshot = snapshot;
-                    _lastProcessedDocument = document;
+                    _lastProcessedSnapshot = currentSnapshot;
+                    _lastProcessedDocument = currentDocument;
+                    _lastProcessedSyntaxTree = currentSyntaxTree;
                 }
 
                 _reportChangeCancellationSource = new CancellationTokenSource();
                 _notificationService.RegisterNotification(() =>
                     {
                         _workQueue.AssertIsForeground();
-                        ReportChangedSpan(snapshot.GetFullSpan());
+                        ReportChangedSpan(changedSpan);
                     },
                     ReportChangeDelayInMilliseconds,
                     _listener.BeginAsyncOperation("ReportEntireFileChanged"),
                     _reportChangeCancellationSource.Token);
             }
 
+            private async Task<SnapshotSpan> GetChangedSpanAsync(
+                Document previousDocument, Document currentDocument,
+                ITextSnapshot currentSnapshot, CancellationToken cancellationToken)
+            {
+                // Try to defer to the classification service to see if it can find a narrower range to recompute.
+                if (previousDocument != null)
+                {
+                    var classificationService = TryGetClassificationService(currentSnapshot);
+                    if (classificationService != null)
+                    {
+                        var changeRange = await ComputeChangeRangeAsync(
+                            previousDocument, currentDocument, classificationService, cancellationToken).ConfigureAwait(false);
+                        if (changeRange != null)
+                            return currentSnapshot.GetSpan(changeRange.Value.Span.Start, changeRange.Value.NewLength);
+                    }
+                }
+
+                // Couldn't compute a narrower range.  Just the mark the entire file as changed.
+                return currentSnapshot.GetFullSpan();
```

Comment:
we fallback to old behavior if we can't get a narrower region.  this will also be the behavior for F#.  (TS doesn't matter as they don't use our classification system, they use TextMate).